lib/trino-parquet/src/main/java/io/trino/parquet/reader/flat/FlatColumnReader.java

```diff
@@ -0,0 +1,365 @@
+        }
+
+        Block block = createNonNullBlock(nextBatchSize, values);
+        return new ColumnChunk(block, EMPTY_DEFINITION_LEVELS, EMPTY_REPETITION_LEVELS);
+    }
+
+    /**
+     * Finds the number of values to be skipped in the current page to reach
+     * the start of the current row range and uses that to skip ValueDecoder
+     * and DefinitionLevelDecoder to the appropriate position.
+     *
+     * @return Whether to skip the entire remaining page
+     */
+    private boolean skipToRowRangesStart()
+    {
+        int skipCount = toIntExact(rowRanges.skipToRangeStart());
+        if (skipCount >= remainingPageValueCount) {
+            remainingPageValueCount = 0;
+            return true;
+        }
+        if (skipCount > 0) {
+            int nonNullsCount;
+            if (field.isRequired()) {
+                nonNullsCount = skipCount;
+            }
+            else {
+                nonNullsCount = definitionLevelDecoder.skip(skipCount);
+            }
+            valueDecoder.skip(nonNullsCount);
+            remainingPageValueCount -= skipCount;
+        }
+        return false;
+    }
+
+    private boolean readNextPage()
+    {
+        DataPage page = pageReader.readPage();
+        if (page == null) {
+            return false;
+        }
+
+        if (page instanceof DataPageV1) {
+            readFlatPageV1((DataPageV1) page);
+        }
+        else if (page instanceof DataPageV2) {
+            readFlatPageV2((DataPageV2) page);
+        }
+
+        remainingPageValueCount = page.getValueCount();
+        rowRanges.resetForNewPage(page.getFirstRowIndex());
+        return true;
+    }
+
+    private void readFlatPageV1(DataPageV1 page)
+    {
+        Slice buffer = page.getSlice();
+        ParquetEncoding definitionEncoding = page.getDefinitionLevelEncoding();
+
+        checkArgument(field.isRequired() || definitionEncoding == RLE, "Invalid definition level encoding: " + definitionEncoding);
+        int alreadyRead = 0;
+        if (definitionEncoding == RLE) {
+            // Definition levels are skipped from file when the max definition level is 0 as the bit-width required to store them is 0.
+            // This can happen for non-null (required) fields or nullable fields where all values are null.
+            // See org.apache.parquet.column.Encoding.getValuesReader for RLE for reference.
+            if (field.getDescriptor().getMaxDefinitionLevel() == 0) {
+                definitionLevelDecoder = new ZeroDefinitionLevelDecoder();
+            }
+            else {
+                int bufferSize = buffer.getInt(0); //  We need to read the size even if nulls are absent
+                definitionLevelDecoder = new FlatDefinitionLevelDecoder(buffer.slice(Integer.BYTES, bufferSize));
+                alreadyRead = bufferSize + Integer.BYTES;
+            }
+        }
+
+        valueDecoder = decodersProvider.create(page.getValueEncoding(), field, dictionary);
```

Comment:
Do these value decoders produce dictionary-encoded blocks or do they resolve the indirection and produce flat blocks?

Reply:
At this point we rely on DictionaryReader to decode the dictionary and use ApacheParquetValueDecoder to process values in batches.
Subsequent PRs will change this to produce dictionary-encoded blocks, I didn't do it here to keep changes small.